lib/puppet/provider/dsc_base_provider/dsc_base_provider.rb

```diff
@@ -665,16 +665,17 @@ def mandatory_set_attributes(context)
     context.type.attributes.select { |_attribute, properties| properties[:mandatory_for_set] }.keys
   end
 
-  # Parses the DSC resource type definition to retrieve the names of any attributes which are specifed as required strings
-  # This is used to ensure that any nil values are converted to empty strings to match puppets expecetd value
+  # Parses the DSC resource type definition to retrieve the names of any attributes which are specified as required strings
+  # This is used to ensure that any nil values are converted to empty strings to match puppets expected value
   # @param context [Object] the Puppet runtime context to operate in and send feedback to
   # @param data [Hash] the hash of properties returned from the DSC resource
+  # @param name_hash [Hash] the hash of namevars to be passed as properties to `Invoke-DscResource`
   # @return [Hash] returns a data hash with any nil values converted to empty strings
-  def stringify_nil_attributes(context, data)
-    nil_strings = data.select { |_name, value| value.nil? }.keys
-    string_attrs = context.type.attributes.select { |_name, properties| properties[:type] == 'String' }.keys
-    string_attrs.each do |attribute|
-      data[attribute] = '' if nil_strings.include?(attribute)
+  def stringify_nil_attributes(context, data, name_hash)
+    managed_nil_attributes = data.select { |_name, value| value.nil? }.keys & name_hash.keys
```

Comment:
just thinking more on this.. is the ampersand valid? `The & operator returns an array that contains elements that are unique and present in both arrays.` - this to me doesn't seem right, and likely breaks the existing setup and not all attributes we were stringifying before were namevars.

Reply:
It was my intention to filter only unique in both arrays. 
Maybe I'm wrong on this but to me we have these cases

```
A. Values is string (or enum accepting empty_string) + Values set in manifest is empty_string => Convert nil to string
B. Values is string (or enum accepting empty_string) + Values is not set in manifest (ie undefined) => Return nil (This is were I'm not sure it is the right behavior to have. But in my opinion, it should be nil if puppet awaits a nil return for undefined resources)
C. Values is not string nor enum accepting string => Not concerned

So the the behavior on B. is what may need to be changed
My fear was : if I return empty string on an undefined value, you may change the system in an unintended manner.

(But as I'm writing this, I'm realizing that if you ever set a value manually, the system won't return a nil value. And if you set an empty string, it may as well lead to a perma change 🤔 )

Well I could have been wrong on this indeed. Let me know what you think. I may indeed change the behavior with the arrays.

Reply:
@jordanbreen28, I guess you were not notified by my response. Please let me know what behavior we should seek. I will update the code accordingly.

Reply:
@Clebam apologies for missing this.
I would say for now could we stick to fixing the issue at hand in this PR, and try to leave the existing behaviour unaltered. We can revisit the original behaviour in a separate PR/issue if we find it to not be correct - does that sound ok?

Reply:
@jordanbreen28 ok in fact the name_hash is not useful so I just add to take every nil values from the context and check if string or Enum['']. This time the behavior is the same as before, just checking the enum with empty_string also. 👍